docs/proposals/rebalancing-clusters-across-shards-dynamically.md

```diff
@@ -0,0 +1,144 @@
+---
+title: Neat-enhancement-idea
+authors:
+  - "@ishitasequeira" # Authors' github accounts here.
+sponsors:
+  - TBD        # List all interested parties here.
+reviewers:
+  - "@alexmt"
+  - TBD
+approvers:
+  - "@alexmt"
+  - TBD
+
+creation-date: yyyy-mm-dd
+last-updated: yyyy-mm-dd
+---
+
+# Neat Enhancement Idea
+
+Rebalance clusters across shards automatically on changes to the number of available shards.
+
+
+## Open Questions [optional]
+
+This is where to call out areas of the design that require closure before deciding to implement the
+design.
+
+
+## Summary
+
+Current implementation of sharding uses StatefulSet for the application controller and the goal is to move towards an agile stateless Deployment. Although the application controller does not have any state to preserve, stateful sets were used to get predictable hostnames and the serial number in the hostname was used to get the shard id of a particular instance.  Using StatefulSet has the following limitations:
+
+Any change done to the StatefulSet would cause all the child pods to restart in a serial fashion. This makes scaling up/down of the application controller slow as even existing healthy instances need to be restarted as well.
```

Comment:
We've learned another StatefulSet limitation: it does not re-create evicted pods.  I think it was a mistake to use StatefulSet and moving back to Deployment is a good idea.

Reply:
Is that really how it works? The docs seem to suggest that pods from a StatefulSet are recreated on eviction; 
> If the pods are managed by a [workload](https://kubernetes.io/docs/concepts/workloads/) resource (such as [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) or [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/)) that replaces failed pods, the control plane or kube-controller-manager creates new pods in place of the evicted pods.

https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/

Reply:
Thanks for the clarification! I learned about StatefulSet and eviction from someone else, but I guess it is handled. Please disregard my comment :)